### PR TITLE
Use pool allocator in mapvector

### PR DESCRIPTION
**Pull request by roystgnr**
I've been wanting to create some clever custom data structure here for a
decade, and I still want to in the long term, but it seems that just
using a better allocator is enough to get my new benchmark tests running
2 or 3% faster.

### CODE REVIEW

**Comment by roystgnr** [comment]
Pinging @fdkong in case he wants to try this out too, perhaps on a larger-scale problem.

**Comment by YaqiWang** [comment]
Is this `index_t` only unsigned int? for example, can it be dof_id_type?

**Comment by jwpeterson** [comment]
Is this a `const` issue? The compiler error:
```
static_assert failed "Allocator::value_type must be same type as value_type"
    static_assert((is_same<typename allocator_type::value_type, value_type>::value),
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
According to [here](https://en.cppreference.com/w/cpp/memory/allocator) the `value_type` of `std::allocator<T>` is `T`, but the `value_type` of [map](https://en.cppreference.com/w/cpp/container/map) is `pair<const Key, Val>`. I'm just wondering why some compilers don't seem to care about that.

**Comment by jwpeterson** [comment]
> Is this index_t only unsigned int? for example, can it be dof_id_type?

@YaqiWang `unsigned int` has always been the default template parameter for mapvector, this PR doesn't change that. Whenever we use it, it's specified to be `dof_id_type`, see, e.g. `mesh/distributed_mesh.C`.

File: include/utils/mapvector.h [hunk 1]
```diff
@@ -20,6 +20,9 @@
 #ifndef LIBMESH_MAPVECTOR_H
 #define LIBMESH_MAPVECTOR_H
 
+// libMesh Includes   -------------------------------
+#include <libmesh/pool_allocator.h>
```
Comment (jwpeterson) [inline include/utils/mapvector.h:24]: We usually use the `""` style of include for libmesh headers. I know there are subtle distinctions between `""` and `<>`, any particular reason for using the latter here?

**Comment by roystgnr** [comment]
> Is this index_t only unsigned int? for example, can it be dof_id_type?

In our internal uses it's currently *only* `dof_id_type`; we just use this for `_nodes` and `_elements` in `DistributedMesh`, and naturally use `dof_id_type` for the index in either case.

File: include/utils/mapvector.h (hunk 1 above)
Comment (roystgnr) [reply include/utils/mapvector.h:24]: Whoops!  Originally I was trying to use Boost's pool allocator directly, before I discovered that Ben had already made a wrapper for it, and I guess I wasn't thinking about the quoting type when I changed that.  I'll fix it.

**Comment by roystgnr** [comment]
> I'm just wondering why some compilers don't seem to care about that.

This is the sort of thing Concepts was intended to help fix in C++20, I guess?  Standards (and user code) where it's required to do generic programming with Foo<T>, but where requirements on T may not be explicitly asserted, only implicitly used or unused under the hood?  So I've written broken code, but only libc++ is specifically asserting non-brokenness; libstdc++ (and possibly libc++ aside from the assert) isn't doing anything that requires that assumption.

**Comment by jwpeterson** [comment]
Makes sense. Yet another time when I'm glad we have compile coverage for more than one compiler/std library.

**Review by jwpeterson — APPROVED** [review]

**Comment by lindsayad** [comment]
I cannot compile after this PR with g++ 9.3 and associated libstdc++
```
In file included from /usr/include/c++/9/map:60,
                 from ./include/libmesh/mapvector.h:27,
                 from ./include/libmesh/distributed_mesh.h:27,
                 from /home/lindad/projects/moose/scripts/../libmesh/src/mesh/distributed_mesh.C:21:
/usr/include/c++/9/bits/stl_tree.h: In instantiation of ‘std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::allocator_type std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::get_allocator() const [with _Key = long unsigned int; _Val = std::pair<const long unsigned int, libMesh::Node*>; _KeyOfValue = std::_Select1st<std::pair<const long unsigned int, libMesh::Node*> >; _Compare = std::less<long unsigned int>; _Alloc = libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Node*> >; std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::allocator_type = libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Node*> >]’:
/usr/include/c++/9/bits/stl_map.h:345:16:   required from ‘std::__cxx1998::map<_Key, _Tp, _Compare, _Alloc>::allocator_type std::__cxx1998::map<_Key, _Tp, _Compare, _Alloc>::get_allocator() const [with _Key = long unsigned int; _Tp = libMesh::Node*; _Compare = std::less<long unsigned int>; _Alloc = libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Node*> >; std::__cxx1998::map<_Key, _Tp, _Compare, _Alloc>::allocator_type = libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Node*> >]’
/usr/include/c++/9/debug/safe_container.h:90:37:   required from ‘__gnu_debug::_Safe_container<_SafeContainer, _Alloc, _SafeBase, _IsCxx11AllocatorAware>& __gnu_debug::_Safe_container<_SafeContainer, _Alloc, _SafeBase, _IsCxx11AllocatorAware>::operator=(__gnu_debug::_Safe_container<_SafeContainer, _Alloc, _SafeBase, _IsCxx11AllocatorAware>&&) [with _SafeContainer = std::__debug::map<long unsigned int, libMesh::Node*, std::less<long unsigned int>, libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Node*> > >; _Alloc = libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Node*> >; _SafeBase = __gnu_debug::_Safe_node_sequence; bool _IsCxx11AllocatorAware = true]’
/usr/include/c++/9/debug/map.h:44:11:   required from here
/usr/include/c++/9/bits/stl_tree.h:575:16: error: no matching function for call to ‘libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Node*> >::FastPoolAllocator(const _Node_allocator&)’
  575 |       { return allocator_type(_M_get_Node_allocator()); }
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ./include/libmesh/mapvector.h:24,
                 from ./include/libmesh/distributed_mesh.h:27,
                 from /home/lindad/projects/moose/scripts/../libmesh/src/mesh/distributed_mesh.C:21:
./include/libmesh/pool_allocator.h:128:12: note: candidate: ‘libMesh::FastPoolAllocator<T>::FastPoolAllocator(const libMesh::FastPoolAllocator<T>&) [with T = std::pair<const long unsigned int, libMesh::Node*>]’
  128 |   explicit FastPoolAllocator(const FastPoolAllocator & o) :
      |            ^~~~~~~~~~~~~~~~~
./include/libmesh/pool_allocator.h:128:56: note:   no known conversion for argument 1 from ‘const _Node_allocator’ {aka ‘const libMesh::FastPoolAllocator<std::_Rb_tree_node<std::pair<const long unsigned int, libMesh::Node*> > >’} to ‘const libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Node*> >&’
  128 |   explicit FastPoolAllocator(const FastPoolAllocator & o) :
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~^
./include/libmesh/pool_allocator.h:124:3: note: candidate: ‘libMesh::FastPoolAllocator<T>::FastPoolAllocator() [with T = std::pair<const long unsigned int, libMesh::Node*>]’
  124 |   FastPoolAllocator() :
      |   ^~~~~~~~~~~~~~~~~
./include/libmesh/pool_allocator.h:124:3: note:   candidate expects 0 arguments, 1 provided
In file included from /usr/include/c++/9/map:60,
                 from ./include/libmesh/mapvector.h:27,
                 from ./include/libmesh/distributed_mesh.h:27,
                 from /home/lindad/projects/moose/scripts/../libmesh/src/mesh/distributed_mesh.C:21:
/usr/include/c++/9/bits/stl_tree.h: In instantiation of ‘std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::allocator_type std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::get_allocator() const [with _Key = long unsigned int; _Val = std::pair<const long unsigned int, libMesh::Elem*>; _KeyOfValue = std::_Select1st<std::pair<const long unsigned int, libMesh::Elem*> >; _Compare = std::less<long unsigned int>; _Alloc = libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Elem*> >; std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::allocator_type = libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Elem*> >]’:
/usr/include/c++/9/bits/stl_map.h:345:16:   required from ‘std::__cxx1998::map<_Key, _Tp, _Compare, _Alloc>::allocator_type std::__cxx1998::map<_Key, _Tp, _Compare, _Alloc>::get_allocator() const [with _Key = long unsigned int; _Tp = libMesh::Elem*; _Compare = std::less<long unsigned int>; _Alloc = libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Elem*> >; std::__cxx1998::map<_Key, _Tp, _Compare, _Alloc>::allocator_type = libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Elem*> >]’
/usr/include/c++/9/debug/safe_container.h:90:37:   required from ‘__gnu_debug::_Safe_container<_SafeContainer, _Alloc, _SafeBase, _IsCxx11AllocatorAware>& __gnu_debug::_Safe_container<_SafeContainer, _Alloc, _SafeBase, _IsCxx11AllocatorAware>::operator=(__gnu_debug::_Safe_container<_SafeContainer, _Alloc, _SafeBase, _IsCxx11AllocatorAware>&&) [with _SafeContainer = std::__debug::map<long unsigned int, libMesh::Elem*, std::less<long unsigned int>, libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Elem*> > >; _Alloc = libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Elem*> >; _SafeBase = __gnu_debug::_Safe_node_sequence; bool _IsCxx11AllocatorAware = true]’
/usr/include/c++/9/debug/map.h:44:11:   required from here
/usr/include/c++/9/bits/stl_tree.h:575:16: error: no matching function for call to ‘libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Elem*> >::FastPoolAllocator(const _Node_allocator&)’
  575 |       { return allocator_type(_M_get_Node_allocator()); }
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ./include/libmesh/mapvector.h:24,
                 from ./include/libmesh/distributed_mesh.h:27,
                 from /home/lindad/projects/moose/scripts/../libmesh/src/mesh/distributed_mesh.C:21:
./include/libmesh/pool_allocator.h:128:12: note: candidate: ‘libMesh::FastPoolAllocator<T>::FastPoolAllocator(const libMesh::FastPoolAllocator<T>&) [with T = std::pair<const long unsigned int, libMesh::Elem*>]’
  128 |   explicit FastPoolAllocator(const FastPoolAllocator & o) :
      |            ^~~~~~~~~~~~~~~~~
./include/libmesh/pool_allocator.h:128:56: note:   no known conversion for argument 1 from ‘const _Node_allocator’ {aka ‘const libMesh::FastPoolAllocator<std::_Rb_tree_node<std::pair<const long unsigned int, libMesh::Elem*> > >’} to ‘const libMesh::FastPoolAllocator<std::pair<const long unsigned int, libMesh::Elem*> >&’
  128 |   explicit FastPoolAllocator(const FastPoolAllocator & o) :
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~^
./include/libmesh/pool_allocator.h:124:3: note: candidate: ‘libMesh::FastPoolAllocator<T>::FastPoolAllocator() [with T = std::pair<const long unsigned int, libMesh::Elem*>]’
  124 |   FastPoolAllocator() :
      |   ^~~~~~~~~~~~~~~~~
./include/libmesh/pool_allocator.h:124:3: note:   candidate expects 0 arguments, 1 provided
```

**Pull request merged**